### PR TITLE
Refac effectiveness matching

### PR DESCRIPTION
**Pull request by davidsantiagoquevedo**
This PR introduces many breaking changes in Vaccineff. The number of functions and steps for computing vaccine effectiveness has been reduced. The new pipeline for estimation consists of three main functions:

1. `make_immunization()`: This function prepares the information on immunization date and vaccine status. It can handle more than one column for vaccine dates and custom vaccine statuses. In such cases, it returns the name of the column that was selected as immunizing and the custom name, if provided.

2. `match_cohort()`: We improved and generalized this function to reduce observation bias in cohorts. The default matching strategy is `static`, based on `nearest` and `exact` characteristics using `Mahalanobis` distance. The exposure times of the pairs are adjusted after matching. In future releases, rolling calendar matching will be introduced as a more accurate method to account for exposure times. The function returns an `s3 object` of type `match`, from which a summary and balance of the cohorts can be printed using the `summary()` method, and the matched cohort can be extracted using the `dataset()` method. The matched cohort contains all the necessary information to estimate vaccine effectiveness.

3. `effectiveness()`: This function receives a (matched) cohort and estimates vaccine effectiveness using the Hazard Ratio (HR). An `s3 object` of type `effectiveness` is returned, compatible with the `plot` and `summary` methods. Future releases will provide relative risk (RR) as an alternative for cases where the proportional hazards hypothesis is not satisfied.

**Related issues**
* The output from `summary(effectiveness)` includes the suggestions made in issue https://github.com/epiverse-trace/vaccineff/issues/49
* The generalized matching method (`match_cohort`) solves issue https://github.com/epiverse-trace/vaccineff/issues/37

**Compatibility**: Users can still access the previous estimation pipeline of Vaccineff via release `0.0.1`.

### CODE REVIEW

**Review by GeraldineGomez — APPROVED** [review]

File: R/coh_match_utils.R [hunk 1]
```diff
@@ -0,0 +1,189 @@
+#' @title Match Cohort Using Mahalanobis Distance
+#'
+#' @description This function constructs pairs of vaccinated and unvaccinated
+#' individuals with similar characteristics. It relies on the matching
+#' algorithm implemented in the `{MatchIt}` package.
+#' By default, the function uses `method = "nearest"`, `ratio = 1`, and
+#' `distance = "mahalanobis"` to perform the matching.
+#' @inheritParams match_cohort
+#' @return A data frame with the matched population.
```
[the remark this answers is not in the document]
Comment (davidsantiagoquevedo) [reply R/coh_match_utils.R:9]: Accepted

File: R/coh_match.R [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+#' @param nearest Named vector with name(s) of column(s) for `nearest` matching
+#' and caliper(s) for each variable (e.g., `nearest = c("characteristic1" = n1,
+#' "characteristic2" = n2)`, where `n1` and `n2` are the calipers). Default is
+#' `NULL`.
+#' @param immunization_date_col Name of the column that contains the
+#' immunization date to set the beginning of the follow-up period
+#' (`t0_follow_up`). Default is `immunization_date`.
+#' @param vacc_status_col name of the column containing the vaccination.
+#' Default is `vaccine_status`.
+#' @param vaccinated_status string assigned to the vaccinated population.
```
[the remark this answers is not in the document]
Comment (davidsantiagoquevedo) [reply R/coh_match.R:43]: Replaced to Status

File: R/coh_immunization.R [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+  # Limit date
+  # If censoring date provided, use it. If not, use outcome date
+  limit_date <- data[[outcome_date_col]]
+
+  if (!is.null(censoring_date_col)) {
+    limit_date <- as.Date(ifelse(!is.na(data[[censoring_date_col]]),
+      yes = as.character(data[[censoring_date_col]]),
+      no = as.character(limit_date)
+    ))
```
[the remark this answers is not in the document]
Comment (davidsantiagoquevedo) [reply R/coh_immunization.R]: I get this error, I think we actually need it to be vectorized to work out
```
Error in if (!is.na(data_set[[censoring_date_col]])) { : 
  the condition has length > 1
```

File: R/coh_match.R [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  # input checking
+  checkmate::assert_data_frame(
+    data,
+    min.rows = 1, min.cols = 1
+  )
+
+  checkmate::assert_names(
+    names(data),
+    must.include = c(outcome_date_col, immunization_date_col, vacc_status_col)
+  )
```
[the remark this answers is not in the document]
Comment (davidsantiagoquevedo) [reply R/coh_match.R]: Good suggestion. I will address this in a future PR where the final structure of the functions is completely defined.

**Pull request merged**